src/imitation/data/types.py

```diff
@@ -69,26 +69,53 @@ def __post_init__(self):
 
 
 @dataclasses.dataclass(frozen=True)
-class Transitions:
-    """A batch of obs-act-obs-done transitions.
+class TransitionsMinimal:
+    """A batch of obs-act transitions.
 
-    Often generated by combining and processing several TrajectoryNoRew objects via
-    `flatten_trajectories()`.
+    This class and its subclasses are usually instantiated via
+    `imitation.data.rollout.flatten_trajectories`.
     """
 
     obs: np.ndarray
     """
     Previous observations. Shape: (batch_size, ) + observation_shape.
 
     The i'th observation `obs[i]` in this array is the observation seen
-    by the agent when choosing action `acts[i]`.
-
-    Invariant: obs.dtype == next_obs.dtype.
+    by the agent when choosing action `acts[i]`. `obs[i]` is not required to
+    be from the timestep preceding `obs[i+1]`.
     """
 
     acts: np.ndarray
     """Actions. Shape: (batch_size,) + action_shape."""
 
+    infos: np.ndarray
+    """Optional array of info dicts. Shape: (batch_size,)."""
+
+    def __len__(self):
+        """Returns number of transitions. Always positive."""
+        return len(self.obs)
+
+    def __post_init__(self):
+        """Performs input validation: check shapes & dtypes match docstring."""
+        if len(self.obs) != len(self.acts):
+            raise ValueError(
+                "obs and acts must have same number of timesteps: "
+                f"{len(self.obs)} != {len(self.acts)}"
+            )
+        if len(self.obs) == 0:
+            raise ValueError("Must have non-zero number of observations.")
+
+        if self.infos is not None and len(self.infos) != len(self.obs):
+            raise ValueError(
+                "obs and infos must have same number of timesteps: "
+                f"{len(self.obs)} != {len(self.infos)}"
+            )
+
+
+@dataclasses.dataclass(frozen=True)
+class Transitions(TransitionsMinimal):
```

Comment:
I am a little weirded out that `TransitionsMinimal` does not contain `dones`, but I guess it makes sense that `next_obs` and `dones` would go together (since it's something you get *after* a transition).

Reply:
I didn't try to come up with a principled reason to split up the fields between classes.

I stuck to what `BC` needed, and then added in `infos` afterwards because I'd need that for a MineRL-variant of `FrameStackTransformation` later on, and maybe other wrappers / datasets would want to stick metadata into all types of transition.